internal/jujuapi/streamproxy.go

```diff
@@ -0,0 +1,150 @@
+// Copyright 2024 Canonical.
+package jujuapi
+
+import (
+	"context"
+	"database/sql"
+	"fmt"
+	"net/http"
+
+	"github.com/gorilla/websocket"
+	"github.com/juju/juju/api/base"
+	jujuparams "github.com/juju/juju/rpc/params"
+	"github.com/juju/names/v5"
+	"github.com/juju/zaputil/zapctx"
+	"go.uber.org/zap"
+
+	"github.com/canonical/jimm/v3/internal/auth"
+	"github.com/canonical/jimm/v3/internal/dbmodel"
+	"github.com/canonical/jimm/v3/internal/errors"
+	"github.com/canonical/jimm/v3/internal/jimmhttp"
+	"github.com/canonical/jimm/v3/internal/openfga"
+)
+
+// A streamProxier serves all HTTP endpoints by proxying
+// messages between the controller and client.
+type streamProxier struct {
+	// TODO(Kian): Refactor the apiServer to use the JIMM API rather than a concrete struct
+	// then we can write unit tests for the stream proxier.
+	apiServer
+}
+
+// Authenticate implements WSServer.Authenticate
+// It attempts to perform basic auth and will return an unauthorized error if auth fails.
+func (s streamProxier) Authenticate(ctx context.Context, w http.ResponseWriter, req *http.Request) (context.Context, error) {
+	_, password, ok := req.BasicAuth()
+	if !ok {
+		return ctx, errors.E(errors.CodeUnauthorized, "authentication missing")
+	}
+	jwtToken, err := s.jimm.OAuthAuthenticator.VerifySessionToken(password)
+	if err != nil {
+		return ctx, errors.E(errors.CodeUnauthorized, err)
+	}
+	email := jwtToken.Subject()
+	ctx = auth.ContextWithSessionIdentity(ctx, email)
+	return ctx, nil
+}
+
+// ServeWS implements jimmhttp.WSServer.
+func (s streamProxier) ServeWS(ctx context.Context, clientConn *websocket.Conn) {
+	writeError := func(msg string, code errors.Code) {
+		var errResult jujuparams.ErrorResult
+		errResult.Error = &jujuparams.Error{
+			Message: msg,
+			Code:    string(code),
+		}
+		err := clientConn.WriteJSON(errResult)
+		if err != nil {
+			zapctx.Error(ctx, "failed to write error message to client", zap.Error(err), zap.Any("client message", errResult))
+		}
+	}
+	user, err := s.jimm.UserLogin(ctx, auth.SessionIdentityFromContext(ctx))
+	if err != nil {
```

Comment:
Nit: you don't tend to leave any space between lines, i.e.,:
```go
f, err := something()
if err != nil {
}
l, err := something2()
if err != nil {
}
g, err := something3()
if err != nil {
}
```
and it can be kinda difficult to read when blocked together like this.

Where as:
```go
f, err := something()
if err != nil {
}

l, err := something2()
if err != nil {
}

g, err := something3()
if err != nil {
}
```

Reads a bit easier.

Reply:
Will change.

Reply:
Thanks, I don't mean to be picky it just makes things a bit difficult if we don't have a bit of cushion between things aha.